File: examples.py
# defining a function in python everything is based on tabs so no curly braces in this case for the function
def debug(tag, outputVal): 
    print(f'{tag} : {outputVal}')    

# variables and data types  just like js you dont really need to define the data type 
x = 5 #int 
y = 2.5 #float
debug('value of x', y)
# booleans so with bools in py you capitalize the first letter  
is_student = True
debug('is this a student', is_student)

# basic math operations 
sum = x+y
dif = x-y
product = x*y
quotient = x/y
debug('sum', sum)
debug('dif', dif)
debug('product', product)
# to prevent from floating type casting a floating point number decimal to int 
debug('qou',int(quotient))

# modulo 
remainder = 13%5
debug('remainder', remainder)

# casting boolean to string 
is_student_str = str(is_student)

debug('is_student_str', is_student_str)

# determine data types 
debug('type of x', type(x))

""" Control statement """

# if statement
age =90
if age >= 18 and age < 65 :
    print('you are an adult ')
elif age >= 65 :
    print('you are a senior citizen')       
else :print('you are a minor')    
    
#  using and/or 
is_sunny = True
is_weekend = True

if  is_sunny and is_weekend :
    print("let's go to the beach")
elif is_sunny or is_weekend :    
    print("no work to do today")
else : 
    print("let's go to work")   

# comparison operators 
a = 10 
b = 20

if a > b :
    # to pass in var you use the f string 
    print(f'{a} is greater than {b}')
      
if a < b :
    print(f'{a} is less than {b}')    

if a == b :
    print(f'{a} is equal to {b}')

z= 5 
is_greater_than_or_equal = z >= 5
debug('is_greater_than_or_equal', is_greater_than_or_equal)

# suppose to do less than butttt nope i get it 
k = 5.0 # floating point number 
is_not_equal = k != 5 # false because the computer knows its still a 5
debug('is_not_equal', is_not_equal)

# inverting booleans  will usually be used in a toggle mode 
is_teacher = True
is_teacher_not = not is_teacher

debug('is_teacher_not (inverted value)', is_teacher_not)

""" DATA STRUCTURES """
# lists (they are an ordered data structure they can contain duplicates ) their content can be modified 
my_list = ['a', 'b', 'c'] # although these are arrays we call them lists in python

# append an item to the list 
my_list.append('d')
debug('appended item in list', my_list)
debug('my_list', my_list[0])
# the last item the difference in comparison with js 
debug('my_list', my_list[-1])

# get range of items 
my_names = ['Bob', 'sam', 'charles', 'arnold']
debug('get range', my_names[1:])

# remove duplicates from list
my_list.append('a')
debug('updated list', my_list)
# set doesnt allow duplicates but we want it as a list again so we wrap the set with a list 
unique_list = list(set(my_list) )
debug('unique list',unique_list)

# tuples (they are ordered and can contain duplicates ). Their contents cannot be modified not square brackets but round brackets for tuples
my_tuple = (1,2,3,4,5)
debug('first item of tuple ', my_tuple[0])
debug('second to last item of tuple ', my_tuple[-2])


# remove dupes from tuple
my_tuple2 = (1,2,3,3,4,5)

unique_tuple = tuple(set(my_tuple2))
debug('unique tuple', unique_tuple)


# Dictionaries (they are unordered and can contain duplicates). Their contents can be modified their keys are unique and their values can be duplicated the work like objects 
my_dict = {'name': 'bob' , 'age':18}
debug('name', my_dict['name'])
debug('age', my_dict['age'])
# modifying 
my_dict['name'] = 'same '
debug('updated name', my_dict['name'])
# adding a new key value pair
my_dict['job'] = 'developer'
debug('job', my_dict['job'])

# remove key value pair 
del my_dict['job']
debug('my dict', my_dict)

""" LOOPS """
# for loops  more like for in loops 
numbers = [0,1,2,4,7,34,3543,123,657]

for number in numbers:
    debug('for loop number', number)
    if number > 30 and number <= 300:
        print('number is greater than 30 and less than or equal to  300')
    elif number > 300 :
        print('number is greater than 300')
    else:
        print('number is less than 30')      


# while loops 
count = 0
while count < 5 :
 debug('while loop count', count)
 count +=1


""" FUNCTIONS AND MODULES"""

# functions with no parameters 
def sayHello():
    print('hello')
    
    
sayHello()


# functions with parameters 
def sayHello2(name):
    print(f'hello {name.capitalize()}') # capitalizing the name 
       
sayHello2('bob')    
sayHello2('chris')    
sayHello2('richmann')    



# modules getting the math func 
import math 

debug('square root ', int(math.sqrt(25)))


""" Classes """

# create class
class Dog:
    # this here is not compulsory in py tho
    # name = ''
    # age = None  #in py we use None not null 
    
    # initializing some of the properties this object wants to use 
    # constructors this takes in params first the self meaning the scope meaning the dog class and after what ever you want to add 
     def __init__(self, name, age):
        self.name = name
        self.age = age    
    # in py you always have to pass self as the first argument even if there are no other args  so py knows it is related to this class
     def bark(self):
        print('woof') 
        
    
my_dog = Dog('bob', 5) 
debug('dog name', my_dog.name)
# calling the method 
my_dog.bark()

"""File handling"""
filename = 'example.txt'

# writing to a file 
# the open function allows you to write the file 
file = open(filename, 'w') # w means write so only write to this 
file.write('hello world this is a text file \n hjgdgshjfkdshfgvdskhvffhysvfskdj \njbdhasgdhdfgshavdgshdvgshcvghdvhvshjsdgfvjh \n djygfydskghafgagfyuyjgfihakfhfj\n ') 
file.close()


# reading entire file contents 
# with for the async sense 
with open(filename, 'r') as file:  # r for read
     content = file.read()
     debug('content', content)


# read file contents line by line 
with open(filename, 'r') as file: 
     for line in file: 
         print('single line :', line.strip())
         
#edit and append to file 
with open(filename, 'a') as file: # a as append
     file.write('\n this is a new line')

""" EXCEPTION HANDLERS """
# handle exceptions where file might not exist
try:
 with open('file_does_not_exist.txt', 'r') as file:   
     content = file.read()
     print(content)
except FileNotFoundError: #an exception that is raised when a file is not found / but you can decide not to specify and catch all
     print('file not found')  

# handle divide by 0 exceptions 



































# lets go the extra mile 
# def testFunc( num: int ):
#     if num > 2 :
#         print('give me a number = 2 or less tahn 2')
#     else :
#         return print('this is what i want ')    
# testFunc(3)
# testFunc(1)
# testFunc(2)
# testFunc('t')

File: examples.js
// main function to be used for console logs
function debug(tag, outputVal) {
  console.log(`${tag} : ${outputVal}`);
}
// variables
var x = 5; //int
var y = 2.5;
debug("value of x", x);

// booleans
var is_student = true;
debug("is this a student ", is_student);

// basic math operators
var sum, dif, product, quotient, remainder;
sum = x + y;
dif = x - y;
product = x * y;
quotient = x / y;
debug("sum", sum);
debug("dif", dif);
debug("product", product);
// # to prevent from floating type casting a floating point number decimal to int
debug("qou", parseInt(quotient));

// modulo
remainder = 13 % 5;
debug("remainder", remainder);

// casting boolean to string
var this_to_string = true;
debug("is this a string", String(this_to_string));

// determine data types
var theStringType = "the string type";
debug("type of this is ", typeof theStringType);

/* if statement */
let age = 90;
if (age >= 18 && age < 65) console.log("you are an adult");
else if (age >= 65) console.log("you are a senior citizen");
else console.log("you are a minor");

// using and / or
var is_sunny, is_weekend;
is_sunny = true;
is_weekend = true;

if (is_sunny && is_weekend) console.log("lets go to the beach");
else if (is_sunny || is_weekend) console.log("no work today");
else console.log("let's go to work");

//  comparison operators
var a, b, z, k;
a = 10;
b = 20;

if (a > b) console.log(`${a} is greater than ${b}`);
if (a < b) console.log(`${a} is less than ${b}`);
if (a === b) console.log(`${a} is equal to ${b}`);

z = 5;
var is_greater_than_or_equal = z >= 5;
debug("is_greater_than_or_equal", is_greater_than_or_equal);

k = 5.0; // floating point
var is_not_equal = k != 5;
debug("is not equal", is_not_equal);

// inverting bools usually used in a toggle
var is_teacher = true;
var is_teacher_not = !is_teacher;

debug("is_teacher_not (inverted value),", is_teacher_not);

/*DATA STRUCTURES*/
// ARRAYS
let my_list = [1, 2, 3, 4, 5, 6, 7, 8, 8, 2, 2, 5];

// append an item to the array
my_list.push(9);

debug("appended item in list", my_list);
debug("my_list", my_list[0]);
// the last item the difference in comparison with js
debug("my_list", my_list[my_list.length - 1]);

// removing duplicates from arrays
var unique_array = [...new Set(my_list)];
debug("unique list", unique_array);

// tuples in js
var myTuple = [1, "hello", true, true];

var unique_tuple = [...new Set(myTuple)];

debug("unique Tuple", unique_tuple);

// objects in js
const my_object = { name: "bob", age: 18 };
debug("name", my_object.name);
debug("age", my_object["age"]);
// modifying
my_object["name"] = "chris";

// adding a new key-value pair
my_object.job = "fake life developer";
debug("job", my_object["job"]);

// remove key value pair
delete my_object.job;
debug("my object ", my_object.name);
console.log(my_object);

// for loops
numbers = [0, 1, 2, 4, 7, 34, 3543, 123, 657];

for (number of numbers) {
  debug("number ", number);
  if (number > 30 && number <= 300)
    console.log("number is greater than 30 or les than or equal to 300 ");
  else if (number > 300) console.log("number is greater than 300");
  else console.log("number is less than 30");
}
// while loop
let count = 0;
while (count < 5) {
  debug("while loop count", count);
  count++;
}

/*FUNCTIONS AND MODULES */

// functions without parameters
function sayHello() {
  console.log("hello");
}

sayHello();

// functions with params
function sayHello2(name) {
  console.log(`hello ${name.charAt(0).toUpperCase() + name.slice(1)}`);
}

sayHello2('yaro')
sayHello2('chris')    
sayHello2('richmann')  

/* CLASSES*/

class Dog {
  constructor(name, age){
    this.name = name;
    this.age = age;
  }

  bark(){
    console.log('woof')
  }
}

var my_dog = new Dog('fido', 3)
debug('my dog', my_dog.name)
my_dog.bark()

/*FILE HANDLING */
const fs = require('fs');
const filename = 'example2.txt'
let data = 'hello world this is a text file \n hjgdgshjfkdshfgvdskhvffhysvfskdj \njbdhasgdhdfgshavdgshdvgshcvghdvhvshjsdgfvjh \n djygfydskghafgagfyuyjgfihakfhfj\n '

// writing to the file synch
fs.writeFileSync(filename, data)


// reading the file asynch 
// way 1
fs.readFile(filename, 'utf-8', (err, data)=>{
  if (err) throw console.log(err);
  console.log(data)

})
//way 2
const asyncfs = require('fs/promises') 
async function readFile(){
  try {
    const data = await asyncfs.readFile(filename, 'utf-8')
    console.log(data)
  } catch (error) {
    console.log(error)
  }
}
readFile()

// reading a single line 
const readline = require('readline'); 
const file = readline.createInterface({ 
  input: fs.createReadStream(filename), 
  output: process.stdout, 
  terminal: false
}); 

file.on('line', (line) => { 
  console.log(line); 
});
